src/librustdoc/visit_ast.rs

```diff
@@ -337,4 +352,58 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
             om.foreigns.push((item, renamed));
         }
     }
+
+    /// Run a few checks on the `#[doc]` attributes for an item:
+    /// - warn if a crate-level attribute is used on a non-`crate` item
+    /// - warn if `#[doc(inline)]`/`#[doc(no_inline)]` is used on a non-`use` item
+    /// - error if conflicting `#[doc(inline)]`/`#[doc(no_inline)]` are found
+    fn check_item_attrs(&mut self, attrs: &'tcx [ast::Attribute], is_crate: bool, is_use: bool) {
+        let mut specified_inline = None;
+
+        for attr in attrs.lists(sym::doc) {
+            // Check for crate-level attributes used on non-crate item
+            if !is_crate
+                && attr
+                    .meta_item()
+                    .map_or(false, |meta| CRATE_LEVEL_ATTRS.iter().any(|attr| meta.path == *attr))
+            {
+                self.cx
+                    .sess()
+                    .struct_span_warn(attr.meta_item().unwrap().span, "this attribute can only be applied at the crate level")
```

Comment:
Maybe add a new error code for this?

Reply:
I was thinking about adding a lint as I said in the original post. Which would be prefered?

Reply:
lints sound fine to me as long as there's something clear rustdoc can do. Any case that's ambiguous should be a hard error I think - so e.g. `#[doc(inline)] struct S;` should be a warning but
```rust
#[doc(inline)]
#[doc(no_inline)]
pub use crate::S;
```
should be a hard error.

Reply:
So you agree with the categorization in warns/errs that I have implemented? The attributes that `rustdoc` can ignore only get a warn, but `#[doc(inline)] #[doc(no_inline)] struct S;` is a hard error.

Reply:
Yes, that sounds good to me :)

Reply:
Here's an example of adding a lint: https://github.com/rust-lang/rust/pull/77856/files